### ISSUE TITLE
Running directly from `cargo test`

### ISSUE THREAD

**Issue by Ekleog** [ipfs-rust/netsim-embed#30]
Hello,

First, once again thank you for netsim-embed! I'm trying to upgrade to 0.7, and am hitting an issue I can't figure out.

AFAICT, `NetworkBuilder` was renamed to `Netsim`. I was using it in ways similar to this:
```rust
    futures::executor::block_on(async move {
        let mut net = Netsim::<String, String>::new(Ipv4Range::local_subnet_10());

        let last_recipient = net.spawn_machine(|_, mut evt| async move {
            // ... spawn a machine that's ready to receive emails and triggers evt once the email was received
        });

        let kannader_server = net.spawn_machine(move |_, _| async move {
            // spawn a machine that's ready to forward emails
            kannader::run(&opt, shutdown).expect("Failed to run kannader");
        });

        let _initial_client = net.spawn_machine(move |_, _| async move {
            // ... spawn a machine that sends an email
        });

        net.spawn().machine(0).recv().await.unwrap();
    });
```

However, with 0.7, it seems to me like we can only spawn `Command`s any longer with `spawn_machine`, and no longer direct closures.

Do you know where I should look at in order to recover behavior similar to the previous `spawn_machine`?

**Comment by dvc94ch** [comment]
ah yes. this functionality was removed because I ran into issues with global variables. I don't recall the details but the decision was made that it would be more reliable to run each machine in it's own process (instead of trying to run it in a thread). I guess if you didn't have any "reliability issues" then sticking with the old version is fine

**Comment by Ekleog** [comment]
Hmm ok thank you for your answer! Just… so I'm not sure I understand, how should I refactor my code so that each machine runs in its own process? I can probably figure out a way to stop relying on mpsc channels so that each machine is a distinct process, but I'm not sure how I'm supposed to orchestrate building the executables so that `cargo test` Just Works™ :/

**Comment by dvc94ch** [comment]
been a while, I'd have to read the sources again. I'd start by understanding how the examples work [0]

- [0] https://github.com/ipfs-rust/netsim-embed/blob/master/.github/workflows/ci.yml#L52

**Comment by Ekleog** [comment]
Hmm ok so it's what I was fearing from reading the examples. It seems to me it's basically not possible to have `cargo test` just work and it requires manually compiling executables before using `netsim-embed`.

Here is an idea to make this possible again. What if the resulting code for the user-land looked like this (pseudo-code):
```rust
fn foo() {
    // ...
}

fn bar() {
    // ...
}

pub fn main() {
    netsim_embed::dispatch_args!(foo, bar);
    netsim_embed::unshare_user();
    let args = libtest_mimic::Arguments::from_args();
    libtest_mimic::run_tests(&args, ...);
}

fn one_test() {
    let s = Netsim::new();
    s.spawn_machine("foo");
    s.spawn_machine("bar");
}
```

Netsim would then run `{current_exe()} --netsim-embed-internal-arg=foo` to spawn the machine `foo`, and `dispatch_args!` would then call `foo` in the process spawned this way.

WDYT? That said I must say I don't know if/when I'd be able to implement it, so I'll probably roll back to 0.6 for now :sweat_smile:

**Comment by Ekleog** [comment]
Actually even better: `spawn_machine` and `dispatch_args!` could collaborate to setup `ipc-channel`-based communication between the two processes with minimal overhead I think. Does that seem like a sensible API to you / would you accept a PR implementing it?

**Comment by dvc94ch** [comment]
@rkuhn is maintaining it at the moment. I don't have any attachment to netsim myself anymore

**Comment by rkuhn** [comment]
A more complete example is [`ipfs-embed`](https://github.com/ipfs-rust/ipfs-embed/tree/master/harness). The idea is to create one executable and test instances of that with a virtual network — which is a bit of extra work of course.

Your idea @Ekleog only works with one module per (test) crate, right? (since there can be only one `main()`) In that sense I find it a little too magical, but maybe I’m not yet fully understanding your proposal. I agree that it would be nice to have a test case that basically says “here are the N behaviours, spawn them as M processes, then run this code to steer the network”. One crucial requirement from my own usage of netsim-embed is that the steering procedure needs to communicate bidirectionally with the behaviours under test, which the ipfs-embed harness does using stdin/stdout and a simple text protocol.

**Comment by Ekleog** [comment]
Well, to give an example of what the API could look like with `ipc-channel` being fully integrated:

```rust
#[derive(serde::Deserialize, serde::Serialize)]
enum MessagesWithMainProcess {
    HereIsAChannel(IpcSender<usize>,
    // ...
}

fn behavior1(c: IpcReceiver<MessagesWithMainProcess>) {
    // ...
}

fn behavior2(c: IpcReceiver<MessagesWithMainProcess>) {
    // ...
}

fn behavior3(c: IpcReceiver<MessagesWithMainProcess>) {
    // ...
}

fn test1() {
    let s = Netsim::new();
    let sender1: IpcSender<MessagesWithMainProcess> = s.spawn_machine("behavior1");
    let sender2: IpcSender<MessagesWithMainProcess> = s.spawn_machine("behavior2");
    // send stuff to sender1 and sender2, communicate over messages sent over channels sent themselves over them, etc.
}

fn test2() {
    let s = Netsim::new();
    let sender1: IpcSender<MessagesWithMainProcess> = s.spawn_machine("behavior1");
    let sender2: IpcSender<MessagesWithMainProcess> = s.spawn_machine("behavior3");
    // send stuff to sender1 and sender2, communicate over messages sent over channels sent themselves over them, etc.
}

fn main() {
    netsim_embed::dispatch_args!(behavior1, behavior2, behavior3);
    netsim_embed::unshare_user();
    let args = libtest_mimic::Arguments::from_args();
    libtest_mimic::run_tests(&args, ...); // dispatches to either test1, test2 or the two; libtest-mimic itself is a pretty nice crate in conjunction with netsim-embed
}
```

The `dispatch_args` would be a no-op if no `--netsim-embed-internals` arg is passed. If `--netsim-embed-internals=behaviorX` is passed, then it sets up the `IpcOneShotServer` to create the channel, calls `behaviorX`, then return.

Then, `spawn_machine` would just spawn `{argv0} --netsim-embed-internals=behaviorX`. It would then handle the other side of creating `IpcOneShotServer`, then return.

Does that make sense? I think this way ipfs-embed's use case could also become much cleaner, and the implementation should be relatively straightforward :)

**Comment by rkuhn** [comment]
How would a channel be serialised/deserialised by serde? This feels like a much bigger project and quite orthogonal to the purpose of netsim-embed — which is about a simulated network after all.

As far as I can see these are the challenges:

- we need an executable to be tested; this cannot be the `test` executable since that one has a magically conjured `main()`
- the test executable needs to offer IPC — most portable is probably stdin/stdout and text (e.g. JSON)
- then we can provide generic test infrastructure for running such an executable and setting up the plumbing to interact with it

My knowledge of cargo may not be complete, so let me know if I’m missing something: producing a separate and dedicated executable needs an explicit and separate entry point. This means source code, which could potentially be generated by a build script, but it also will be specific to the program under test — the IPC commands need to be hooked up to the actual functions that are to be tested. So to me it seems that the test author will at least have to create a new “frontend” for whatever they’re testing, and explain that to cargo. Generic infrastructure can then invoke cargo and run the executable when spawning a machine for a test, and it can also serialise/deserialise message over the stdin/stdout channel (adding sub-channel multiplexing seems like overkill to me here).

With this preamble, the `main` you sketch should dispatch to one of the behaviours (not to a test!), but this could be made generic, a one-liner in `build.rs`. The behaviours need to be available in the synthetic executable’s crate (e.g. via a dependency to the main crate — but then they need to be visible in the default config, not just `#[cfg(test)]`). Is this what you meant?

**Comment by Ekleog** [comment]
> How would a channel be serialised/deserialised by serde?

According to ipc-channel's readme, it's already implemented :) https://crates.io/crates/ipc-channel

> we need an executable to be tested; this cannot be the test executable since that one has a magically conjured main()

Assuming you use libtest-mimic (which needs to happen anyway as unshare_user must happen before any thread gets spawned), this is not a problem, and can be written in a way similar to the one I gave in my example above — it just requires setting harness = false in the cargo.toml :)

> the test executable needs to offer IPC — most portable is probably stdin/stdout and text (e.g. JSON)

So using ipc-channel could make this happen nicely I think

> producing a separate and dedicated executable needs an explicit and separate entry point

Exactly, this is what I don't know how to do currently (to be precise, I think this is about artifact dependencies, which is WIP on the cargo side). But I think that with the design above we don't need to do that, as we could reuse the same test executable

----

As I'm not sure I'm explaining it correctly, here is my end-goal:
> `cargo test` should be able to seamlessly run multiple tests all defined in a single rust test file, without any need for the user to compile anything ahead-of-time

Do you agree with this goal, and think you could accept reasonable PRs doing it? If so, I'll try to write a PR implementing this with an example, to make what I mean more concrete :)

**Comment by rkuhn** [comment]
This required some reading on my part: exactly how tests are compiled and run was one of the few things about Rust that I hadn’t yet investigated. So the plan is to offer library support for writing a dedicated test under `tests` such that

- the executable does all needed setup like `unshare_user`
- then runs a sequence of tests, declared manually by the author (because `#[test]` is magic, i.e. we cannot use it)
- offers a nice API for starting netsim with a bunch of nodes whose behaviour is defined as simple functions in the same file and sets up ipc-channels with them; API could accept behaviours as `fn(IpcReceiver<T>, IpcSender<T>)` and we look up the symbol’s name to pass it to the child process — this keeps it type-safe
 
That could work and it does sound like a nice improvement! Fleshing this out with an example is a great first step, so please go ahead — and please excuse my initial slowness.

**Comment by Ekleog** [comment]
I've opened https://github.com/ipfs-rust/netsim-embed/pull/31 with the suggested improvement's implementation along with an example integration test that both checks this keeps working and acts as an example for how to use it :)
